api/types/ui_config.go

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright 2023 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package types
+
+import (
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// UIConfig defines configuration for the web UI served
+// by the proxy service. This is a configuration resource,
+// never create more than one instance of it.
+type UIConfig interface {
+	Resource
+
+	// GetScript returns the contents of the installer script
+	GetScrollbackLines() int32
+	// SetScript sets the installer script
+	SetScrollbackLines(int32)
+
+	String() string
+}
+
+func NewUIConfigV1() (*UIConfigV1, error) {
+	uiconfig := &UIConfigV1{
+		ResourceHeader: ResourceHeader{},
+	}
+	if err := uiconfig.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return uiconfig, nil
+}
+
+// CheckAndSetDefaults verifies the constraints for UIConfig.
+func (c *UIConfigV1) CheckAndSetDefaults() error {
+	c.setStaticFields()
+	if err := c.Metadata.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
```

Comment:
Should you be checking the validity of the `c.ScrollbackLines` here?

I assume it has to be > 0, but does xterm.js enforce any other maximum or minimum values?

Reply:
Looks like xterm.js will go all the way up to max u32 (which we can't support unless you change from int32 to uint32), but that's not a major issue.

https://github.com/xtermjs/xterm.js/blob/9154831afa9aea78306d8edad8fe403616758ad3/src/common/services/OptionsService.ts#L170-L173

Reply:
Since xtermjs _only_ throws an error if its less than zero then I'll do that validation in `CheckAndSetDefaults`. I think keeping _our max_ at 2bil lines is probably viable for now. I am prepared for someone to ask for 4bil lines next week because of this decision.

Reply:
> I am prepared for someone to ask for 4bil lines next week because of this decision.

🤣